From 58be23549ccea5f0d2a48fccf612d00bc48523d7 Mon Sep 17 00:00:00 2001
From: Vlad Zahorodnii <vlad.zahorodnii@kde.org>
Date: Mon, 7 Jul 2025 12:27:31 +0300
Subject: [PATCH] xwayland: Forbid flips if toplevel window's winSize is empty

If there is a parent window and a child window with different drawable
sizes but empty winSize (because of XShape) and the child presents a
pixmap, xwl_present_flip() should not be called because that can change
the size of the wl_surface so it no longer matches the size of the
toplevel window.

It fixes input breaking when playing video games in KDE Plasma and GNOME
Shell.

Video games tend to create helper override redirect windows in the top
left screen corner. Due to xwl_present_flip() getting unexpectedly
getting called, those helper windows can have very big wl_surfaces,
which then steal pointer input from regular applications.

Signed-off-by: Vlad Zahorodnii <vlad.zahorodnii@kde.org>
---
 hw/xwayland/xwayland-present.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/hw/xwayland/xwayland-present.c b/hw/xwayland/xwayland-present.c
index 07d58ee6a2..8d16a9c154 100644
--- a/hw/xwayland/xwayland-present.c
+++ b/hw/xwayland/xwayland-present.c
@@ -769,6 +769,9 @@ xwl_present_check_flip(RRCrtcPtr crtc,
     if (!xwl_pixmap_get_wl_buffer(pixmap))
         return FALSE;
 
+    if (!RegionNotEmpty(&toplvl_window->winSize))
+        return FALSE;
+
     /* Window must be same region as toplevel window */
     if ( !RegionEqual(&present_window->winSize, &toplvl_window->winSize) )
         return FALSE;
-- 
GitLab

